Interleave TCP/TPP better

visheshh opened the pull request:
<!--- Please review your changes in preview mode -->
<!--- Provide a general summary of your changes in the Title above -->

#### Bug/feature Description
Interleave TCP/TPP better
#### Affected Platform(s)
All

#### Cause / Analysis / Design
Requests from front end commands, scheduler, server, and moms need to be treated at a different priority level to ensure that the high priority events are handled sooner instead of waiting on lower priority events.

RPP messages at times can block commands and scheduler run commands until the server has processed all of the ones in the queue as do_rpp is called infinitely for a rpp_request

#### Solution Description
check for the active scheduler sockets and do a select read operation on the sockets to process them first .
do_rpp is only called 3 times for a rpp_request

#### Testing logs/output
[Test logs.txt](https://github.com/PBSPro/pbspro/files/2518542/Test.logs.txt)
[multiple_sockets_test.txt](https://github.com/PBSPro/pbspro/files/2518911/multiple_sockets_test.txt)


#### Checklist:
<!--- Use the preview button to see the checkboxes/links properly. -->
<!--- Go over the following points, and put an `x` (without spaces around it) in the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have joined the **[pbspro community forum](http://community.pbspro.org/)**.
- [x] My pull request contains a **single, signed** commit. See **[setting up gpg signature](https://pbspro.atlassian.net/wiki/display/DG/Signing+Your+Git+Commits)**.
- [x] My code follows the **[coding style](https://pbspro.atlassian.net/wiki/display/DG/Coding+Standards)** of this project.
- [ ] My change requires project documentation. See **[required documentation checklist](https://pbspro.atlassian.net/wiki/display/DG/Checklist+for+Developing+Features+and+Bug+Fixes)** for details.
   - [ ] I have added documentation in the **[project documentation area](https://pbspro.atlassian.net/wiki/display/PD)**.
- [x] I have added new **PTL test(s) to my commit**. (See **[using PTL for testing](https://pbspro.atlassian.net/wiki/display/DG/Using+PTL+for+Testing)**) *(or)*
   - [ ] I have added  **manual test(s) to this pull request and explained why PTL is not appropriate** for this case.
- [ ] All new and existing automated tests have passed. (See **[running automated PTL tests](https://pbspro.atlassian.net/wiki/display/DG/PTL+Quick+Start+Guide)**).
- [x] I have attached **test logs to this pull request** as evidence of testing/verification.


__***For further information please visit the [Developer Guide Home](https://pbspro.atlassian.net/wiki/display/DG/Developer+Guide+Home).***__

suresh-thelkar commented on src/include/net_connect.h:
```diff
@@ -172,6 +172,10 @@ enum conn_type {
 	Idle
 };
 
+struct sched_socks {
```
Better to have a generic name for this structure as net_server.c is part of the library(Libnet).

visheshh replied:
done

suresh-thelkar commented on src/lib/Libnet/net_server.c:
```diff
@@ -89,7 +89,8 @@
 static int conns_array_size = 0;  /* Size of the svr_conn list, initialized to 0 */
-
+pbs_sched	*psched;
+pbs_list_head	svr_allscheds;
```
Remove these variables and define wherever required. As this code is part of Libnet library we need to keep generic variables and functions only.

visheshh replied:
done

suresh-thelkar commented on src/lib/Libnet/net_server.c:
```diff
@@ -497,10 +532,9 @@ engage_authentication(conn_t *pconn)
  */
 int
-wait_request(time_t waittime)
+wait_request(time_t waittime, struct sched_socks scks)
```
As said earlier please use generic name instead of struct sched_socks.

suresh-thelkar replied:
Also you can just pass a pointer to the structure. This we can just pass NULL while second parameter is not applicable.

visheshh replied:
done

suresh-thelkar commented on src/lib/Libnet/net_server.c:
```diff
@@ -549,26 +605,7 @@ wait_request(time_t waittime)
-			}
-
-			svr_conn[idx]->cn_func(svr_conn[idx]->cn_sock);
+			process_socket(em_fd);
```
Validate if process_socket() returns failure.

suresh-thelkar replied:
Yes. You can log a message at debug level.

visheshh replied:
done

suresh-thelkar commented on src/resmom/mom_main.c:
```diff
@@ -6143,20 +6144,21 @@ do_rpp(int stream)
  *
  */
-
+#define MAX_RPP_LOOPS 3
```
Just try to put some comment telling why we retry for 3 times.

visheshh replied:
done

visheshh replied:
done

suresh-thelkar commented on src/server/failover.c:
```diff
@@ -119,6 +119,7 @@ extern char	     server_host[];
 
 extern struct connection *svr_conn;
 extern struct batch_request *saved_takeover_req;
+struct sched_socks scks = {0};
```
No need to initialise the structure early.

visheshh replied:
done

suresh-thelkar commented on src/server/pbsd_main.c:
```diff
@@ -2143,11 +2145,28 @@ main(int argc, char **argv)
+				scks.active_socks = scks.active_socks + 1;
+			}
+		}
+		scks.socket_fd = (int *)calloc(scks.active_socks, sizeof(int));
```
Need to validate the failure case of calloc.

suresh-thelkar replied:
Also no need to typecast for calloc.

visheshh replied:
done

suresh-thelkar commented on test/tests/functional/pbs_multi_sched.py:
```diff
@@ -1990,3 +1990,41 @@ def test_multi_sched_node_sort_key(self):
+        t = int(time.time())
+        self.server.manager(MGR_CMD_SET, SCHED,
+                            {'scheduling': 'True'}, id='sc3', expect=True)
+        self.server.log_match("processing priority sockets", starttime=t)
```
Test looks good to me. No need of third scheduler here. Two are enough.

visheshh replied:
done

visheshh commented:
@suresh-thelkar , I have addressed  review comments , Please review

suresh-thelkar commented on src/lib/Libnet/net_server.c:
```diff
@@ -472,6 +471,39 @@ engage_authentication(conn_t *pconn)
+ *
+ */
+static int
+process_socket(int sock)
```
Better to keep the name of the function as process_priority_socket only.

visheshh replied:
It is a common function to process all the fd's .

suresh-thelkar replied:
Now that you are using this function for normal as well as priority sockets no need to rename it.

visheshh replied:
done

suresh-thelkar commented on src/lib/Libnet/net_server.c:
```diff
@@ -485,6 +517,7 @@ engage_authentication(conn_t *pconn)
  *	routine associated with the socket is invoked.
  *
  * @param[in] waittime - Timeout for tpp_em_wait (poll)
+ * @param[in] scks - structure contains active fds and count
```
Could you elaborate more on the purpose of introducing this structure. For example you can say something like if this structure is populated then we can only process the set of sockets present only in this set as a priority, else if this parameter is NULL then we process all sockets.

visheshh replied:
done

suresh-thelkar commented on src/lib/Libnet/net_server.c:
```diff
@@ -528,6 +560,28 @@ wait_request(time_t waittime)
+				 	if (FD_ISSET(scks->socket_fd[i], &fdset)) {
+						log_event(PBSEVENT_DEBUG3, PBS_EVENTCLASS_SERVER,
+							LOG_DEBUG, __func__, "processing priority sockets");
+						return process_socket(scks->socket_fd[i]);
```
We need to return from this function after processing all the priority sockets. But the current code returns as soon as it processes only one socket from the list of priority sockets. Please change the code accordingly.

visheshh replied:
done

suresh-thelkar commented on src/server/pbsd_main.c:
```diff
@@ -2143,11 +2145,37 @@ main(int argc, char **argv)
 			reap_child();
 #endif	/* WIN32 */
-
+		scks = (struct priority_socks *)malloc(sizeof(struct priority_socks));
```
Instead of allocating memory dynamically using two function calls one at this location and the other at line no #2161 you can actually club these two function calls into one if we allocate memory at the line no #2161 itself. Because by that time we actually know how much memory is indeed required.

visheshh replied:
removed this malloc

suresh-thelkar commented on src/resmom/mom_main.c:
```diff
@@ -6143,21 +6143,18 @@ do_rpp(int stream)
-	int	i;
-	/* To reduce rpp process storm reducing max do_rpp processing to 3 times */
-	for (i=0 ; i < MAX_RPP_LOOPS ; i++) {
+	for (;;) {
```
Have you removed this code temporarily? Hope you are adding back this code again.

visheshh replied:
Added back

subhasisb commented:
@visheshh as discussed we should not create two different contexts. That will be wrong. If we do that we will be waiting separately for 2 different types of requests (front-end - i.e. qsub/qstats etc and scheduler requests) - that would artificially increase our overall wait time. We need to do the exact opposite, we need to wait in one single atomic call for everything (all sockets and signals) which is achieved by the tpp_em_pwait() call in wait_request(). To give priority to the scheduler commands, we then find out whether a scheduler connection has data by doing an additional select/poll (important point is that it should be with ZERO timeout) to detect that and then process those socket first.

suresh-thelkar commented on src/lib/Libnet/net_server.c:
```diff
@@ -528,47 +561,64 @@ wait_request(time_t waittime)
+						log_event(PBSEVENT_DEBUG3, PBS_EVENTCLASS_SERVER,
+							LOG_DEBUG, __func__, "processing priority sockets");
+						process_socket(scks->socket_fd[i]);
+						sched_processed = 1;
```
Better to rename the variable sched_processed to something like prty_socks_processed for better readability. Priority sockets can contain any other sockets apart from Scheduler fds in future.

visheshh replied:
done

visheshh replied:
done

suresh-thelkar commented on src/lib/Libnet/net_server.c:
```diff
@@ -528,47 +561,64 @@ wait_request(time_t waittime)
+			if (scks && scks->active_socks && sched_processed) {
+				for (j = 0; j < scks->active_socks; j++) {
+                                	if (scks->socket_fd[i] == em_fd) {
+                                        	sched_sock = 1;
```
No need to use the variable sched_sock to find out whether the current em_fd is already processed or not. We can use the following check to get the desired result after the above loop.
if (j != scks->active_socks) ==> em_fd is already processed.
else em_fd is not processed.

visheshh replied:
done

visheshh commented:
@suresh-thelkar , I have addressed your comments , Please review.

suresh-thelkar commented on src/lib/Libnet/net_server.c:
```diff
@@ -549,26 +608,20 @@ wait_request(time_t waittime)
-			if (idx < 0) {
-				return -1;
+			j = 0;
+			if (scks && scks->active_socks && prio_sock_processed) {
```
You can change the condition as follows.
if (scks && prio_sock_processed && scks->active_socks && )

visheshh replied:
If there are not active sockets at all there is no need to check if its even processed . So keeping it the same . I'll change the next condition.

suresh-thelkar commented on src/lib/Libnet/net_server.c:
```diff
@@ -549,26 +608,20 @@ wait_request(time_t waittime)
-						continue;
-					}
-				}
+			if (scks && (j < scks->active_socks) && prio_sock_processed) {
```
Here also you can change the condition as follows.
if (scks && prio_sock_processed &&(j < scks->active_socks))

visheshh replied:
done

visheshh commented on src/include/net_connect.h:
```diff
@@ -172,6 +172,10 @@ enum conn_type {
 
+struct priority_socks {
+        int *socket_fd;	/*Array of active socket file descriptors*/
+        int active_socks;	/*Number of active scheduler sockers*/
```
done

visheshh commented on src/include/net_connect.h:
```diff
@@ -172,6 +172,10 @@ enum conn_type {
 	Idle
 };
 
+struct priority_socks {
```
done

visheshh commented on src/lib/Libnet/net_server.c:
```diff
@@ -497,15 +531,14 @@ engage_authentication(conn_t *pconn)
 {
-	int nfds;
-	int idx;
+	int nfds,i,j;
```
done

visheshh commented on src/lib/Libnet/net_server.c:
```diff
@@ -497,15 +531,14 @@ engage_authentication(conn_t *pconn)
+	int nfds,i,j;
 	em_event_t *events;
-	int err,i;
+	int err,sched_processed,sched_sock;
```
done

visheshh commented on src/lib/Libnet/net_server.c:
```diff
@@ -528,47 +561,64 @@ wait_request(time_t waittime)
+			timeout.tv_sec = 0;
+			FD_ZERO(&fdset);
+                        for (i = 0; i < scks->active_socks; i++) {
+                                if (scks->socket_fd[i] != 0) {
```
done

visheshh commented on src/server/pbsd_main.c:
```diff
@@ -2142,11 +2144,31 @@ main(int argc, char **argv)
+		/* assign the active sockets to socket_fd */
+		for (psched = (pbs_sched*) GET_NEXT(svr_allscheds); psched; psched = (pbs_sched*) GET_NEXT(psched->sc_link)) {
+			if (psched->scheduler_sock != -1) {
+				scks.socket_fd[act_scks]=psched->scheduler_sock;
```
done

visheshh commented:
@suresh-thelkar Added priority flag to connection structure as discussed to optimize check for scheduler fd as discussed. Please review

visheshh commented:
@subhasisb @suresh-thelkar , Please review the updated changes.

subhasisb commented on src/lib/Libnet/net_server.c:
```diff
@@ -610,17 +615,12 @@ wait_request(time_t waittime, priority_socks *scks)
 #endif
 			if(prio_sock_processed) {
```
please add a space after if

visheshh replied:
done

subhasisb commented on src/lib/Libnet/net_server.c:
```diff
@@ -1072,6 +1117,11 @@ init_poll_context(void)
 
 #endif
 	DBPRT(("#init_poll_context: initializing poll_context for %d", max_connection))
+	priority_context = tpp_em_init(max_connection);
```
A very minor nit. Since you are doing the tpp_em_add_fd of poll_context first and priority_context next, initializing in that order would be nicer to look at

visheshh replied:
done

subhasisb commented on src/lib/Libnet/net_server.c:
```diff
@@ -98,7 +98,8 @@ int	max_connection = -1;
 static int	net_is_initialized = 0;
 static void	*poll_context;  /* This is the context of the descriptors being polled */
-static int	init_poll_context();  /* Initialized the poll context */
+void 	*priority_context;
```
Can you also initialize this to NULL

visheshh replied:
done

subhasisb commented on src/server/run_sched.c:
```diff
@@ -302,8 +302,7 @@ contact_sched(int cmd, char *jobid, pbs_net_t pbs_scheduler_addr, unsigned int p
 #endif
-
-	static int priority_flag = 1;
+	const int priority_flag = 1;
```
We usually use a macro for such things instead of a constant

visheshh replied:
done

visheshh commented:
@subhasisb , Please review the updated changes.

subhasisb commented:
LGTM @visheshh - please do remember to rerun your tests since a lot has changed through the review.

visheshh commented on src/include/net_connect.h:
```diff
@@ -211,6 +213,7 @@ struct connection {
 	time_t		cn_lasttime;	/* time last active */
 	void		(*cn_func)(int); /* read function when data rdy */
 	void		(*cn_oncl)(int); /* func to call on close */
+	int		cn_prio_flag;	/* flag for a priority socket */
```
done

visheshh commented on src/lib/Libnet/net_server.c:
```diff
@@ -99,7 +98,8 @@ int	max_connection = -1;
 static int	net_is_initialized = 0;
 static void	*poll_context;  /* This is the context of the descriptors being polled */
-static int	init_poll_context();  /* Initialized the poll context */
+void 	*priority_context = NULL;
```
done

visheshh commented on src/lib/Libnet/net_server.c:
```diff
@@ -520,14 +556,42 @@ wait_request(time_t waittime)
+			em_event_t *pevents;
+			timeout = 0;
+#ifndef WIN32
+        		/* wait after unblocking signals in an atomic call */
```
removed errno set to 0 for Windows as it is already done as part of function definition

visheshh commented on src/lib/Libnet/net_server.c:
```diff
@@ -520,14 +556,42 @@ wait_request(time_t waittime)
+        		err = errno;
+#endif /* WIN32 */
+                	for (i = 0; i < pnfds; i++) {
+                        	int em_pfd;
```
done

bhroam replied:
I'm not sure I agree.  Unless there is a reason to do otherwise, I think variables should be as tightly scoped as possible.  There won't be a runtime difference since variable declarations will be handled at compile time.

visheshh commented on src/lib/Libnet/net_server.c:
```diff
@@ -693,11 +772,22 @@ add_conn(int sd, enum conn_type type, pbs_net_t addr, unsigned int port, void (*
+                	snprintf(logbuf, sizeof(logbuf),
+                        	"could not add socket %d to the priority poll list", sd);
+                	log_err(err, __func__, logbuf);
+               		close_conn(sd);
```
done

visheshh commented on src/lib/Libnet/net_server.c:
```diff
@@ -1039,6 +1144,14 @@ init_poll_context(void)
+	if ((tpp_em_add_fd(priority_context, sd_dummy, EM_IN) == -1)) {
+		int err = errno;
+		snprintf(logbuf, sizeof(logbuf),
+			"Could not add socket %d to the read set", sd_dummy);
```
done

suresh-thelkar commented on src/lib/Libnet/net_server.c:
```diff
@@ -592,10 +591,9 @@ wait_request(time_t waittime, void *priority_context)
 			}
 		}
 
+		int em_fd;
```
As discussed this variable needs to be defined at the start of the function otherwise Windows build might mail.

visheshh replied:
done

visheshh commented:
[windows test logs.txt](https://github.com/PBSPro/pbspro/files/2695364/windows.test.logs.txt)
[failover.txt](https://github.com/PBSPro/pbspro/files/2695367/failover.txt)
[Test logs- multisched.txt](https://github.com/PBSPro/pbspro/files/2695368/Test.logs-.multisched.txt)

Attaching test logs . Squashed and rebased.
